libglusterfs/src/event-epoll.c

```diff
@@ -73,73 +82,77 @@ __event_slot_alloc(struct event_pool *event_pool, int fd,
     int j = 0;
     int table_idx = 0;
     int gen;
-    struct event_slot_epoll *table = NULL;
+    struct event_slot_epoll *table_slot;
+    struct event_slot_epoll_table *table = NULL;
 
 retry:
 
     while (table_idx < EVENT_EPOLL_TABLES) {
         table = event_pool->ereg[table_idx];
         if (table) {
-            if (table->slots_used == EVENT_EPOLL_SLOTS) {
+            if (table->table_slots_used == EVENT_EPOLL_SLOTS) {
                 table_idx++;
                 continue;
             } else {
-                break; /* break out of the loop */
+                goto search_slot; /* break out of the loop */
             }
         } else {
             table = __event_newtable(event_pool, table_idx);
             if (!table) {
                 return -1;
             } else {
-                break;
+                table_slot = &table->slots[0];
+                gen = 0;
+                goto set_slot;
             }
         }
     }
 
-    if (table_idx >= EVENT_EPOLL_TABLES)
-        return -1;
+    // (table_idx >= EVENT_EPOLL_TABLES) - not more tables
+    return -1;
 
+search_slot:
     for (j = 0; j < EVENT_EPOLL_SLOTS; j++) {
-        if (table[j].fd == -1) {
+        table_slot = &table->slots[j];
+        if (table_slot->fd < 0) {
             /* wipe everything except bump the generation */
-            gen = table[j].gen;
-            memset(&table[j], 0, sizeof(table[j]));
-            table[j].fd = fd;
-            table[j].gen = gen + 1;
-
-            INIT_LIST_HEAD(&table[j].poller_death);
-            LOCK_INIT(&table[j].lock);
-
-            if (notify_poller_death) {
-                table[j].idx = table_idx * EVENT_EPOLL_SLOTS + j;
-                list_add_tail(&table[j].poller_death,
-                              &event_pool->poller_death);
-            }
-
-            event_pool->ereg[table_idx]->slots_used++;
-
-            break;
+            gen = table_slot->gen;
+            goto set_slot;
         }
     }
 
-    if (j == EVENT_EPOLL_SLOTS) {
-        table = NULL;
-        table_idx++;
-        goto retry;
+    // (j == EVENT_EPOLL_SLOTS) - retry with the next table
+    table = NULL;
+    table_idx++;
+    goto retry;
+
+set_slot:
+    memset(table_slot, 0, sizeof(struct event_slot_epoll));
+    table_slot->fd = fd;
+    table_slot->gen = gen + 1;
+
+    LOCK_INIT(&table_slot->lock);
+
+    if (notify_poller_death) {
+        table_slot->idx = table_idx * EVENT_EPOLL_SLOTS + j;
```

Comment:
`j` may be incorrect here if we have created a new table after going through line 125. However I think it should be impossible to get to line 125 (we should have found an unused entry because `table_slots_used < EVENT_EPOLL_SLOTS`.

Reply:
Just in case, added a 'j = 0' after line 105.
I do think it shouldn't indeed happen, as some slot has to be available.